### DISCUSSION TITLE
Proposal: Enable Self-Supervised Learning in OpenFL

### DISCUSSION THREAD

**Discussion by porteratzo** [securefederatedai/openfederatedlearning#1316] [Ideas]
# Proposal: Enable Self-Supervised Learning in OpenFL

## Summary

Introduce Self-Supervised Learning (SSL) algorithms into OpenFL to enable training on unlabeled data. This can be achieved by creating workflows that utilize techniques such as Masked Autoencoders (MAE) or DinoV2. These algorithms can allow models to learn useful patterns from raw data without needing human-labeled annotations, which can then be fine-tuned on labeled data for specific tasks. The final model is expected to achieve better accuracy than models trained solely on labeled data while reducing dependence on labeled datasets.

## Motivation

Collaborators often face challenges in contributing to federated learning training due to the scarcity of labeled data. Generally, unlabeled data is much more abundant compared to labeled data. However, this resource remains underutilized without a method to train with unlabeled data. With SSL, we can allow collaborators to contribute to training with their unlabeled data, improving the final accuracy of the trained models.

## High-Level Design

The feature will contain two workspaces: one dedicated to SSL pretraining and another for traditional fine-tuning that can utilize the weights from the pre-trained model.

### Dataset
We propose using the BraTS2020 dataset, which is already approved for Intel. The dataset can be accessed [here](https://www.med.upenn.edu/cbica/brats2020/data.html).

### Pretraining Workspace
- **Data_download.sh**: Bash script to download the BraTS2020 dataset.
- **HuggingFace Transformers TaskRunner**: Inherits from `PyTorchTaskRunner` and implements `Transformers Trainer` for an  DinoV2 SSL task for training with OpenFL TaskRunner API.
- **BraTS2020Dataloader**: Inherits `openfl.federated.data.DataLoader` and handles preprocessing (utilizing the entire dataset without labels) and provides the dataloaders the TaskRunner needs.
  - A parameter, `label_uniformity_alpha`, will determine the uniformity of the label distribution among collaborators using [Dirichlet Distribution-Based Partitioning](https://arxiv.org/abs/1905.12022).

### Fine-Tuning Workspace
- **Data_download.sh**: Bash script to download the BraTS2020 dataset.
- **HuggingFace Transformers TaskRunner**: Inherits from `PyTorchTaskRunner` and implements `Transformers Trainer` for a fine-tuning task for training with OpenFL TaskRunner API.
- **BraTS2020Dataloader**: Inherits `openfl.federated.data.DataLoader` and handles preprocessing (utilizing a subset for training with labels) and provides the data loaders the TaskRunner needs.
  - A parameter, `label_uniformity_alpha`, will determine the uniformity of the label distribution among collaborators using [Dirichlet Distribution-Based Partitioning](https://arxiv.org/abs/1905.12022).

### Documentation
- Steps to run pretraining followed by fine-tuning on the pre-trained model.
- Steps to run only fine-tuning with a randomly initialized model.
- Instructions for using `label_uniformity_alpha` to compare Independent and Identically Distributed (IID) data vs. Non-Independent and Identically Distributed (Non-IID) data, where SSL pretraining excels.

### Result Comparison Notebook
- A Jupyter notebook to compare the final results of the fine-tuned model against the pre-trained model.
- This will help users evaluate the effectiveness of the SSL pretraining process and make informed decisions about model performance.

### API Changes
- Introduce new options for `fx workspace create --template.`

### Dependencies
- Deep learning libraries related to training HuggingFace Transformers.

### Backward Compatibility
- The feature will not affect backward compatibility.

## Alternatives Considered
- **Fully Supervised Learning with Synthetic Data Generation**: This approach was not chosen due to the complexity and potential inaccuracies in generating synthetic data.
- **Semi-Supervised Learning**: While effective, it still requires a significant amount of labeled data, which may not always be available.

## Risks and Mitigations
- **Data Privacy Concerns**: Mitigation through secure federated learning.

## Open Questions
- Random initialization vs transfer learning from pre-trained DinoV2
- Any other algorithms preferable to is DinoV2 the most suitable SSL algorithm for the proposed use case, or should other SSL algorithms be considered?
- Support out of domain pretraining? Maybe look for other MRI datasets of different organs and fine-tune on only one organ.
- Include Parameter Efficient Fine-Tuning? It can be much faster to train than training the entire model and vastly more bandwidth efficient.
- Implement the DinoV2 data curation pipeline? If customers want to use DinoV2 SSL, using this pipeline to curate their unlabeled data and remove any noisy/unrepresentative data could be beneficial.
- HuggingFace Transformers has a robust configuration system. Should we consider integrating part of it into OpenFL plan.yaml?

**Comment by kminhta** [comment]
Thanks @porteratzo ! This looks good overall. I think this a strong use case. I particularly like the introduction of a HF transformers task runner. Even beyond this use case, it's a good framework level feature to enable.

I have a couple questions (or open discussion):

1. do you have a sense of the generalizability of these workspaces? I think an issue all the workspaces have is that they end up very nuanced with the intention being that a developer would modify it to their use case. Considering that this proposal will introduce two new workspaces, we'll end up with two "half workspaces" that are intended to be chained together and it begs the question of whether this work would be enabling a new feature or showcasing how to use existing OpenFL features for this use case (minus the HF transformers task runner - that is new)
2. related to point 1, have you considered a design choice that would give a developer the option to toggle between pretraining and finetuning? I think it would better demonstrate self-supervised learning as a unique feature. I don't know how practical that is since pretraining and fine-tuning would fundamentally be separate experiments. Just an open thought
3. can you expand a bit on the `label_uniformity_alpha`? Not necessarily from a technical standpoint. I mean, would this be a parameter for partitioning the data? In other words, this is meant purely for simulation, correct?
4. can you also explain a bit more the distinction between the pretraining and fine-tuning workspaces? Since DinoV2 are foundation models, we would be fine-tuning the backbone for brain MRI images even during the pretraining step. Is the "pretraining step" an unsupervised training of the backbone to learn basic brain MRI features and the "fine-tuning" workspace a supervised training of a specific segmentation head with a frozen backbone?

Anyway, this is interesting! Would be great to be able to enable this in OpenFL
